kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ProtocolMetadataLogReplay.java

```diff
@@ -0,0 +1,207 @@
+import org.slf4j.LoggerFactory;
+
+/** Static utility class for loading Protocol and Metadata from Delta log files. */
+public class ProtocolMetadataLogReplay {
+
+  private static final Logger logger = LoggerFactory.getLogger(ProtocolMetadataLogReplay.class);
+
+  /** Read schema when searching for the latest Protocol and Metadata. */
+  public static final StructType PROTOCOL_METADATA_READ_SCHEMA =
+      new StructType().add("protocol", Protocol.FULL_SCHEMA).add("metaData", Metadata.FULL_SCHEMA);
+
+  /** Result of loading Protocol and Metadata from a LogSegment. */
+  public static class Result {
+    public final Protocol protocol;
+    public final Metadata metadata;
+    private final long logFilesRead;
+
+    public Result(Protocol protocol, Metadata metadata, long logFilesRead) {
+      this.protocol = protocol;
+      this.metadata = metadata;
+      this.logFilesRead = logFilesRead;
+    }
+  }
+
+  /**
+   * Loads the latest Protocol and Metadata from the log files in the given LogSegment.
+   *
+   * <p>Uses the provided lazy CRC loader to bound how many delta files it reads, and to ensure we
+   * only read the CRC file if needed.
+   *
+   * <p>We read delta files in reverse order (newest first) searching for the latest Protocol and
+   * Metadata. When we reach the version just before (greater than) the CRC file version (if
+   * present), we lazily load the CRC file to fill in any missing Protocol or Metadata, avoiding
+   * reading older delta files.
+   *
+   * <p>Also validates that the Kernel can read the table at the loaded protocol version.
+   */
+  public static Result loadProtocolAndMetadata(
+      Engine engine,
+      Path dataPath,
+      LogSegment logSegment,
+      Lazy<Optional<CRCInfo>> lazyCrcInfo,
+      SnapshotMetrics snapshotMetrics) {
+    final Result result =
+        snapshotMetrics.loadProtocolMetadataTotalDurationTimer.time(
+            () -> loadProtocolAndMetadataInternal(engine, logSegment, lazyCrcInfo));
+
+    TableFeatures.validateKernelCanReadTheTable(result.protocol, dataPath.toString());
+
+    logger.info(
+        "[{}] Took {}ms to load Protocol and Metadata at version {}, read {} log files",
+        dataPath,
+        snapshotMetrics.loadProtocolMetadataTotalDurationTimer.totalDurationMs(),
+        logSegment.getVersion(),
+        result.logFilesRead);
+
+    return result;
+  }
+
+  private static Result loadProtocolAndMetadataInternal(
+      Engine engine, LogSegment logSegment, Lazy<Optional<CRCInfo>> lazyCrcInfo) {
+    final long snapshotVersion = logSegment.getVersion();
+    final Optional<FileStatus> crcFileOpt = logSegment.getLastSeenChecksum();
+    final Optional<Long> crcVersionOpt =
+        crcFileOpt.map(f -> FileNames.checksumVersion(f.getPath()));
+
+    // If CRC is at this exact snapshot version, use it directly
+    if (crcVersionOpt.isPresent() && crcVersionOpt.get() == snapshotVersion) {
+      final Optional<CRCInfo> crcInfo = lazyCrcInfo.get();
```

Comment:
should we validate that the provided lazyCrcInfo version matches our expectation from logSegment?